// restful-node/dist/errors/http.error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Generic HttpError, with status code 500
 * */
class HttpError extends Error {
    constructor(message) {
        super(message);
        this.status = 500;
    }
}
exports.HttpError = HttpError;
class NotFoundError extends HttpError {
    constructor(message = null) {
        super(message);
        this.status = 404;
    }
}
exports.NotFoundError = NotFoundError;
class BadRequestError extends HttpError {
    constructor(message) {
        super(message);
        this.status = 400;
    }
}
exports.BadRequestError = BadRequestError;
class ForbiddenError extends HttpError {
    constructor(message) {
        super(message);
        this.status = 403;
    }
}
exports.ForbiddenError = ForbiddenError;
class UnauthorizedError extends HttpError {
    constructor(message) {
        super(message);
        this.status = 401;
    }
}
exports.UnauthorizedError = UnauthorizedError;
class EntityTooLargeError extends HttpError {
    constructor(message) {
        super(message);
        this.status = 413;
    }
}
exports.EntityTooLargeError = EntityTooLargeError;
class MethodNotAllowed extends HttpError {
    constructor(message = 'Method Not Allowed') {
        super(message);
        this.status = 405;
    }
}
exports.MethodNotAllowed = MethodNotAllowed;
